### PR TITLE
Check rewind data is not null [db breaking change, require resync]

### PR DESCRIPTION
**Pull request by dangershony**
There is a race condition where we reason the rewind index and a reorg happen that removed the rewind data from store.

### CODE REVIEW

**Comment by dangershony** [comment]
We might also consider removing items form RewindDataIndex when w rewind operation takes place

File: src/Stratis.Bitcoin.Features.Consensus/Rules/ProvenHeaderRules/ProvenHeaderCoinstakeRule.cs [hunk 1]
```diff
@@ -359,14 +359,17 @@ private UnspentOutputs CheckIfCoinstakeIsSpentOnAnotherChain(ProvenBlockHeader h
             RewindData rewindData = this.PosParent.UtxoSet.GetRewindData(rewindDataIndex.Value).GetAwaiter().GetResult();
 
             UnspentOutputs matchingUnspentUtxo = null;
-            foreach (UnspentOutputs unspent in rewindData.OutputsToRestore)
+            if (rewindData != null)
```
[the remark this answers is not in the document]
Comment (dangershony) [reply src/Stratis.Bitcoin.Features.Consensus/Rules/ProvenHeaderRules/ProvenHeaderCoinstakeRule.cs]: Well good question I guess not, but when we initialise rewind index we make do this checks (perhaps we should throw at that point)

```
            if (rewindData == null)
            {
                throw new ConsensusException($"Rewind data of height '{rewindHeight}' was not found!");
            }

            if (rewindData.OutputsToRestore == null || rewindData.OutputsToRestore.Count == 0)
            {
                return;
            }
```

File: src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/RewindDataIndexCache.cs [hunk 2]
```diff
@@ -29,52 +31,67 @@ public class RewindDataIndexStore : IRewindDataIndexStore
...
-                }
+                this.AddRewindData(rewindHeight, rewindData);
+            }
+        }
 
-                if (rewindData.OutputsToRestore == null || rewindData.OutputsToRestore.Count == 0)
-                {
-                   continue;
-                }
+        private void AddRewindData(int rewindHeight, RewindData rewindData)
```
Comment (noescape00) [inline src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/RewindDataIndexCache.cs]: No logs in the whole class 🙁

**Comment by noescape00** [comment]
Looks correct to me.

**Comment by dangershony** [comment]
This is a breaking change because the rewind data height is now calculated at the correct height and not N-1 as before.

This means resync form scratched

**Pull request merged**